pkg/operator/metricscontroller/metrics_controller.go

```diff
@@ -0,0 +1,109 @@
+package metricscontroller
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	prometheusapi "github.com/prometheus/client_golang/api"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+
+	"k8s.io/client-go/transport"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+type MetricsSyncFunc func(ctx context.Context, controllerContext factory.SyncContext, promClient prometheusv1.API) error
+
+type metricsController struct {
+	operatorClient          v1helpers.OperatorClient
+	prometheusQuery         string
+	newPrometheusClientFunc func() (prometheusv1.API, idleConnectionCloser, error)
+	metricsSyncFunc         MetricsSyncFunc
+}
+
+// NewMetricsController creates a new metrics controller using an in-cluster prometheus client.
+// The given service CA path will be used to read out the CA bundle being trusted by the prometheus client.
+// The controller executes the given metrics sync function every minute.
+func NewMetricsController(operatorClient v1helpers.OperatorClient, recorder events.Recorder, serviceCAPath string, metricsSyncFunc MetricsSyncFunc) factory.Controller {
+	c := &metricsController{
+		operatorClient: operatorClient,
+		newPrometheusClientFunc: func() (prometheusv1.API, idleConnectionCloser, error) {
+			return newInClusterPrometheusClient(serviceCAPath)
+		},
+		metricsSyncFunc: metricsSyncFunc,
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(time.Minute).
+		ToController("MetricsController", recorder.WithComponentSuffix("metrics-controller"))
```

Comment:
need a better name that can be flavored or set by a name argument.

Reply:
Indeed, good suggestion.

Reply:
fixed